File: components/VideoControlsBar.cpp
#include "VideoControlsBar.h"
#include "../style/StyleProps.h"

#include <QHBoxLayout>
#include <QLabel>
#include <QPushButton>

#include <QTimer>
#include <QPalette>
#include <QColor>

namespace {
constexpr qint64 kSeekStepMs = 2000;
}

VideoControlsBar::VideoControlsBar(QWidget* parent): QWidget(parent) {
  buildUi();
  wireSignals();
  setEnabledForMedia(false);
  setPlaying(false);
  setPlaybackRate(1.0);
  setMuted(false);
}

void VideoControlsBar
::buildUi() {
  auto* layout = new QHBoxLayout(this);
  layout->setContentsMargins(0, 0, 0, 0);
  layout->setSpacing(8);

  playButton_       = new QPushButton("Play", this);
  pauseButton_      = new QPushButton("Pause", this);
  backButton_       = new QPushButton("⟵ 2s", this);
  forwardButton_    = new QPushButton("2s ⟶", this);
  slowerButton_     = new QPushButton("Slower", this);
  resetSpeedButton_ = new QPushButton("Reset 1.0x", this);
  fasterButton_     = new QPushButton("Faster", this);
  muteButton_       = new QPushButton("Mute", this);

  Style::setSize(playButton_, "md");
  Style::setSize(pauseButton_, "md");
  Style::setSize(backButton_, "md");
  Style::setSize(forwardButton_, "md");
  Style::setSize(slowerButton_, "md");
  Style::setSize(resetSpeedButton_, "md");
  Style::setSize(fasterButton_, "md");
  Style::setSize(muteButton_, "md");
  
  Style::setVariant(playButton_, "primary");
  Style::setVariant(pauseButton_, "secondary");
  Style::setVariant(backButton_, "outline");
  Style::setVariant(forwardButton_, "outline");
  Style::setVariant(slowerButton_, "ghost");
  Style::setVariant(resetSpeedButton_, "ghost");
  Style::setVariant(fasterButton_, "ghost");
  Style::setVariant(muteButton_, "outline");
  

  
  // Prevent space/enter focus weirdness in analysis workflows
  /*
  If later you want the shadcn-like “focus ring”, you’ll have to allow focus
  and prevent “Enter triggers default button” via setAutoDefault(false).
  */
  playButton_       ->setFocusPolicy(Qt::NoFocus);
  pauseButton_      ->setFocusPolicy(Qt::NoFocus);
  backButton_       ->setFocusPolicy(Qt::NoFocus);
  forwardButton_    ->setFocusPolicy(Qt::NoFocus);
  slowerButton_     ->setFocusPolicy(Qt::NoFocus);
  resetSpeedButton_ ->setFocusPolicy(Qt::NoFocus);
  fasterButton_     ->setFocusPolicy(Qt::NoFocus);
  muteButton_       ->setFocusPolicy(Qt::NoFocus);
  
  muteButton_->setCheckable(true);
  
  speedLabel_ = new QLabel(this);
  Style::setRole(speedLabel_, "muted");
  updateSpeedLabel();

  layout->addWidget(playButton_);
  layout->addWidget(pauseButton_);
  layout->addSpacing(8);
  layout->addWidget(backButton_);
  layout->addWidget(forwardButton_);
  layout->addSpacing(8);
  layout->addWidget(slowerButton_);
  layout->addWidget(resetSpeedButton_);
  layout->addWidget(fasterButton_);
  layout->addStretch(1);
  layout->addWidget(speedLabel_);
  layout->addSpacing(8);
  layout->addWidget(muteButton_);

  // control bar button keyboard hotkeys and tooltips:
  playButton_       ->setToolTip("space  Play");
  pauseButton_      ->setToolTip("space  Pause");
  backButton_       ->setToolTip("⟵  Back");
  forwardButton_    ->setToolTip("⟶  Forward");
  slowerButton_     ->setToolTip("{  Slower");
  fasterButton_     ->setToolTip("}  Faster");
  resetSpeedButton_ ->setToolTip("\\  Reset speed");
}

void VideoControlsBar::wireSignals() {
  // Play / Pause
  connect(playButton_,  &QPushButton::clicked, this, &VideoControlsBar::flashPlayButton);
  connect(playButton_,  &QPushButton::clicked, this, &VideoControlsBar::playRequested);

  connect(pauseButton_, &QPushButton::clicked, this, &VideoControlsBar::flashPauseButton);
  connect(pauseButton_, &QPushButton::clicked, this, &VideoControlsBar::pauseRequested);

  // Seek back / forward
  connect(backButton_,  &QPushButton::clicked, this, &VideoControlsBar::flashSeekBackButton);
  connect(backButton_,  &QPushButton::clicked, this, [this]() { emit seekRequestedMs(-kSeekStepMs); });

  connect(forwardButton_, &QPushButton::clicked, this, &VideoControlsBar::flashSeekForwardButton);
  connect(forwardButton_, &QPushButton::clicked, this, [this]() { emit seekRequestedMs(+kSeekStepMs); });

  // Speed controls
  connect(slowerButton_, &QPushButton::clicked, this, &VideoControlsBar::flashSlowerButton);
  connect(slowerButton_, &QPushButton::clicked, this, &VideoControlsBar::slowerRequested);

  connect(fasterButton_, &QPushButton::clicked, this, &VideoControlsBar::flashFasterButton);
  connect(fasterButton_, &QPushButton::clicked, this, &VideoControlsBar::fasterRequested);

  connect(resetSpeedButton_, &QPushButton::clicked, this, &VideoControlsBar::flashResetSpeedButton);
  connect(resetSpeedButton_, &QPushButton::clicked, this, &VideoControlsBar::resetSpeedRequested);

  // Mute: flash on click, emit state on toggled
  connect(muteButton_, &QPushButton::clicked, this, &VideoControlsBar::flashMuteButton);
  connect(muteButton_, &QPushButton::toggled, this, &VideoControlsBar::muteToggled);
}


void VideoControlsBar::setEnabledForMedia(bool enabled) {
  playButton_->setEnabled(enabled);
  pauseButton_->setEnabled(enabled);
  backButton_->setEnabled(enabled);
  forwardButton_->setEnabled(enabled);
  slowerButton_->setEnabled(enabled);
  fasterButton_->setEnabled(enabled);
  resetSpeedButton_->setEnabled(enabled);
  muteButton_->setEnabled(enabled);
}

void VideoControlsBar::setPlaying(bool playing) {
  playButton_->setEnabled(!playing);
  pauseButton_->setEnabled(playing);
}

void VideoControlsBar::setPlaybackRate(double rate) {
  playbackRate_ = rate;
  updateSpeedLabel();
}

void VideoControlsBar::setMuted(bool muted) {
  muteButton_->blockSignals(true);
  muteButton_->setChecked(muted);
  muteButton_->setText(muted ? "Unmute" : "Mute");
  muteButton_->blockSignals(false);
}

void VideoControlsBar::updateSpeedLabel() {
  speedLabel_->setText(QString("Speed: %1x").arg(playbackRate_, 0, 'f', 2));
}

void VideoControlsBar::flashButtonBorder(QPushButton* button) {
  if (!button) return;

  QTimer* timer = flashTimers_.value(button, nullptr);
  if (!timer) {
    timer = new QTimer(button);
    timer->setSingleShot(true);
    flashTimers_.insert(button, timer);

    connect(timer, &QTimer::timeout, this, [button]() {
      if (!button) return;
      button->setProperty("flash", false);
      button->style()->unpolish(button);
      button->style()->polish(button);
      button->update();
    });
  }

  button->setProperty("flash", true);
  button->style()->unpolish(button);
  button->style()->polish(button);
  button->update();

  timer->start(150);
}






void VideoControlsBar::flashPlayButton()       { flashButtonBorder(playButton_); }
void VideoControlsBar::flashPauseButton()      { flashButtonBorder(pauseButton_); }
void VideoControlsBar::flashSeekBackButton()   { flashButtonBorder(backButton_); }
void VideoControlsBar::flashSeekForwardButton(){ flashButtonBorder(forwardButton_); }
void VideoControlsBar::flashSlowerButton()     { flashButtonBorder(slowerButton_); }
void VideoControlsBar::flashFasterButton()     { flashButtonBorder(fasterButton_); }
void VideoControlsBar::flashResetSpeedButton() { flashButtonBorder(resetSpeedButton_); }
void VideoControlsBar::flashMuteButton()       { flashButtonBorder(muteButton_); }

File: AVA_V02_py/components/video_controls_bar.py
"""Video transport controls used by the AVA player."""

from qt_compat import QtCore, QtWidgets, Signal
from styles import set_role, set_size, set_state, set_variant


class VideoControlsBar(QtWidgets.QWidget):
    play_requested = Signal()
    pause_requested = Signal()
    seek_requested_ms = Signal(int)
    slower_requested = Signal()
    faster_requested = Signal()
    reset_speed_requested = Signal()
    mute_toggled = Signal(bool)
    fullscreen_toggle_requested = Signal()

    def __init__(self, parent: QtWidgets.QWidget | None = None) -> None:
        super().__init__(parent)
        self._playback_rate = 1.0
        self._flash_timers: dict[QtWidgets.QPushButton, QtCore.QTimer] = {}
        self._build_ui()
        self._wire_signals()
        self.set_enabled_for_media(False)
        self.set_playing(False)
        self.set_playback_rate(1.0)
        self.set_muted(False)

    def _build_ui(self) -> None:
        layout = QtWidgets.QHBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(8)

        self.play_button = QtWidgets.QPushButton("Play", self)
        self.pause_button = QtWidgets.QPushButton("Pause", self)
        self.back_button = QtWidgets.QPushButton("⟵ 2s", self)
        self.forward_button = QtWidgets.QPushButton("2s ⟶", self)
        self.slower_button = QtWidgets.QPushButton("Slower", self)
        self.reset_speed_button = QtWidgets.QPushButton("Reset 1.0x", self)
        self.faster_button = QtWidgets.QPushButton("Faster", self)
        self.mute_button = QtWidgets.QPushButton("Mute", self)
        self.fullscreen_button = QtWidgets.QPushButton("Fullscreen", self)
        self.mute_button.setCheckable(True)
        self.fullscreen_button.setCheckable(True)

        for button in (
            self.play_button,
            self.pause_button,
            self.back_button,
            self.forward_button,
            self.slower_button,
            self.reset_speed_button,
            self.faster_button,
            self.mute_button,
            self.fullscreen_button,
        ):
            set_size(button, "md")
            button.setFocusPolicy(QtCore.Qt.FocusPolicy.NoFocus)

        set_variant(self.play_button, "primary")
        set_variant(self.pause_button, "secondary")
        set_variant(self.back_button, "outline")
        set_variant(self.forward_button, "outline")
        set_variant(self.slower_button, "ghost")
        set_variant(self.reset_speed_button, "ghost")
        set_variant(self.faster_button, "ghost")
        set_variant(self.mute_button, "outline")
        set_variant(self.fullscreen_button, "outline")

        self.speed_label = QtWidgets.QLabel(self)
        set_role(self.speed_label, "muted")

        layout.addWidget(self.play_button)
        layout.addWidget(self.pause_button)
        layout.addSpacing(8)
        layout.addWidget(self.back_button)
        layout.addWidget(self.forward_button)
        layout.addSpacing(8)
        layout.addWidget(self.slower_button)
        layout.addWidget(self.reset_speed_button)
        layout.addWidget(self.faster_button)
        layout.addStretch(1)
        layout.addWidget(self.speed_label)
        layout.addSpacing(8)
        layout.addWidget(self.mute_button)
        layout.addWidget(self.fullscreen_button)

        self.play_button.setToolTip("space  Play")
        self.pause_button.setToolTip("space  Pause")
        self.back_button.setToolTip("⟵  Back")
        self.forward_button.setToolTip("⟶  Forward")
        self.slower_button.setToolTip("{  Slower")
        self.faster_button.setToolTip("}  Faster")
        self.reset_speed_button.setToolTip("\\  Reset speed")
        self.mute_button.setToolTip("Mute/Unmute audio")
        self.fullscreen_button.setToolTip("F  Fullscreen")
        self.speed_label.setToolTip("{ / }  Change speed, \\ reset")

    def _wire_signals(self) -> None:
        self.play_button.clicked.connect(self.flash_play_button)
        self.play_button.clicked.connect(self.play_requested)

        self.pause_button.clicked.connect(self.flash_pause_button)
        self.pause_button.clicked.connect(self.pause_requested)

        self.back_button.clicked.connect(self.flash_seek_back_button)
        self.back_button.clicked.connect(lambda: self.seek_requested_ms.emit(-2000))

        self.forward_button.clicked.connect(self.flash_seek_forward_button)
        self.forward_button.clicked.connect(lambda: self.seek_requested_ms.emit(2000))

        self.slower_button.clicked.connect(self.flash_slower_button)
        self.slower_button.clicked.connect(self.slower_requested)

        self.faster_button.clicked.connect(self.flash_faster_button)
        self.faster_button.clicked.connect(self.faster_requested)

        self.reset_speed_button.clicked.connect(self.flash_reset_speed_button)
        self.reset_speed_button.clicked.connect(self.reset_speed_requested)

        self.mute_button.clicked.connect(self.flash_mute_button)
        self.mute_button.toggled.connect(self.mute_toggled)

        self.fullscreen_button.clicked.connect(self.flash_fullscreen_button)
        self.fullscreen_button.clicked.connect(self.fullscreen_toggle_requested)

    def set_enabled_for_media(self, enabled: bool) -> None:
        for button in (
            self.play_button,
            self.pause_button,
            self.back_button,
            self.forward_button,
            self.slower_button,
            self.faster_button,
            self.reset_speed_button,
            self.mute_button,
            self.fullscreen_button,
        ):
            button.setEnabled(enabled)

    def set_playing(self, playing: bool) -> None:
        self.play_button.setEnabled(not playing)
        self.pause_button.setEnabled(playing)

    def set_playback_rate(self, rate: float) -> None:
        self._playback_rate = rate
        self.speed_label.setText(f"Speed: {rate:.2f}x")
        set_state(self.slower_button, "speedActive", rate < 1.0)
        set_state(self.faster_button, "speedActive", rate > 1.0)

    def set_muted(self, muted: bool) -> None:
        self.mute_button.blockSignals(True)
        self.mute_button.setChecked(muted)
        self.mute_button.setText("Unmute" if muted else "Mute")
        self.mute_button.blockSignals(False)

    def set_fullscreen(self, fullscreen: bool) -> None:
        self.fullscreen_button.blockSignals(True)
        self.fullscreen_button.setChecked(fullscreen)
        self.fullscreen_button.setText("Exit Fullscreen" if fullscreen else "Fullscreen")
        self.fullscreen_button.blockSignals(False)

    def _flash_button_border(self, button: QtWidgets.QPushButton) -> None:
        if button not in self._flash_timers:
            timer = QtCore.QTimer(button)
            timer.setSingleShot(True)
            timer.timeout.connect(lambda b=button: set_state(b, "flash", False))
            self._flash_timers[button] = timer

        set_state(button, "flash", True)
        self._flash_timers[button].start(150)

    def flash_play_button(self) -> None:
        self._flash_button_border(self.play_button)

    def flash_pause_button(self) -> None:
        self._flash_button_border(self.pause_button)

    def flash_seek_back_button(self) -> None:
        self._flash_button_border(self.back_button)

    def flash_seek_forward_button(self) -> None:
        self._flash_button_border(self.forward_button)

    def flash_slower_button(self) -> None:
        self._flash_button_border(self.slower_button)

    def flash_faster_button(self) -> None:
        self._flash_button_border(self.faster_button)

    def flash_reset_speed_button(self) -> None:
        self._flash_button_border(self.reset_speed_button)

    def flash_mute_button(self) -> None:
        self._flash_button_border(self.mute_button)

    def flash_fullscreen_button(self) -> None:
        self._flash_button_border(self.fullscreen_button)
